### DISCUSSION TITLE
Why is `additive` true by default?

### DISCUSSION THREAD

**Discussion by noah-nuebling** [Flinesoft/BartyCrouch#257] [Q&A] [answered]
Hi!

First off, thanks @Jeehut for making this, it's super cool.

I'm new to localization, and I'm looking into bartycrouch right now. I was wondering why the `additive` option for the code task is true by default? 

As far as I understands, it prevents cleanup of keys in the .strings file that aren't used in code anymore.
I can't think of good reasons to turn it on. And I thought that suggests that I'm not thinking about some possible scenario where it would be useful.

So what's the rationale behind having it on?

Thank you!

**Comment by Jeehut** [comment, marked as answer]
@noah-nuebling Thank you for your kind words. I'm glad you like BartyCrouch, but if you're new to it, I recommend you switch to my app [ReMafoX](https://github.com/FlineDev/ReMafoX/releases) which is currently in Open Beta and will be released soon. It's like BartyCrouch put into a nice UI and improved with additional features – and obviously easier to use thanks to a UI.

To answer your question: It only makes sense to disable the `additive` option if all your keys are actually going to be available in code `NSLocalizedString` references with static String keys. As soon as you use other references in code, such as the recommended `L10n` method using SwiftGen (see [this blog post](https://jeehut.medium.com/localization-in-swift-like-a-pro-48164203afe2?sk=da26d918390db21261b7ead4837286fc) linked in the README), disabling `additive` would effectively wipe all your keys each time the build script runs.

There are also other popular ways people like to reference their keys with a dynamic string inside the `NSLocalizedString` like [this extension on StackOverflow](https://stackoverflow.com/a/29384360/3451975), which also wouldn't work with `additive` turned off.

If you don't use any of these, you can of course disable `additive`. But disabling it by default would cause trouble for many developers who use dynamic String keys, so it has to be turned on by default for the tool to "just work" in most projects.

**Comment by noah-nuebling** [comment]
@Jeehut thanks very much for the clear explanations and advice.

I would like to check out ReMafoX. The UI looks really good. However, I can't download the beta right now. It says it's not available in my region. I think my App Store might still be set to US.

I don't fully see the need for a UI, since bartycrouch is already has a really nice 'setup once then forget' workflow and the documentation is very good. It took me around a day to get to the point where I feel like I understand how localization works, how to setup and use bartycrouch and what the development and translation workflows are going to be. Using the UI might have shaved a few hours from that time, but now that I have things set up, I don't see a strong need to switch to the UI. It looks really good though. I will definitely keep it in mind as I continue development. 

If there's something else I should be thinking about, I'll definitely appreciate any feedback. Thank you!

**Reply by Jeehut** [reply]
The beta is just [this download link](https://github.com/FlineDev/ReMafoX/releases/download/1.0.0-beta.2/ReMafoX.app.zip), a ZIP file uploaded to the GitHub repo. I'm sure you can download it from everywhere.

The UI of ReMafoX is meant to simplify configuration and explain all the different options, most BartyCrouch users aren't fully aware what they can configure and while documentation isn't that bad at all, people still need to read through it, which takes more time. It also has a UI for adding new keys with review capabilities (see GIF below).

Also ReMafoX has a lot of more to offer, here's a list of **improvments over BartyCrouch** in version 1.0:

- A **new config file format** which can be edited with an intuitive **visual editor** (no `.bartycrouch.toml` file needed)
- A new **"Preview matching files" button** to conveniently simulate a files search & adjust paths accordingly based on results
- An explanation for every single configurable option for more confidence in adjusting the config to your needs
- A **built-in SwiftUI-compatible enum generator** with lots of customization options (no need for SwiftGen for Strings)
- The API keys for machine translations services are no longer stored in the config file (no secrets in Git)
- A **built-in project scanner** to provide smarter defaults for a quick start – including a BartyCrouch migrator (read-only)
- Multiple **built-in step-by-step guides** for easier integration with Xcode
- Adding new localizations **no longer breaks Xcodes edit history** thanks to a built-in Xcode Source Editor extension

Just see this GIF for the improved ReMafoX workflow, you can review all translations & more in this form:

![](https://user-images.githubusercontent.com/6942160/177779457-2e709729-a774-4c60-b8db-1747e84bf74c.gif)

And there's more **planned for the future**:

- Easy way to send translations to translators & integrate their provided translations
- Synchronize translations between Android & iOS platforms of the same app
- Provide a glossary of reviewed common Strings (to improve machine-translated apps)
- UI/Unit tests integration to provide context screenshots for keys for translators
- ... and much more!

I hope that's plenty of reasons to switch to ReMafoX at some point in the future. 😁

**Reply by noah-nuebling** [reply]
Oh thanks for clearing that up. I'm not sure how I ended up in the Mac App Store. Sorry for the confusion. I'll check it out!
